sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/SymbolicExecutionRunner.cs

```diff
@@ -0,0 +1,77 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2020 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Rules.CSharp;
+using SonarAnalyzer.SymbolicExecution;
+
+namespace SonarAnalyzer.Rules.SymbolicExecution
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(EmptyNullableValueAccess.DiagnosticId)]
+    [Rule(ObjectsShouldNotBeDisposedMoreThanOnce.DiagnosticId)]
+    public sealed class SymbolicExecutionRunner : SonarDiagnosticAnalyzer
+    {
+        private readonly SymbolicExecutionAnalyzerFactory symbolicExecutionAnalyzerFactory = new SymbolicExecutionAnalyzerFactory();
```

Comment:
the rules are registered here at lines 35-36 and also in `SymbolicExecutionAnalyzerFactory`

to improve maintainability, I suggest to keep the rules only in this class and pass them to the constructor of `SymbolicExecutionAnalyzerFactory` - and maybe rename that class to something to suggest that is checks if the rules are enabled or not - maybe `EnabledRuleFilter` (it's not really used as a factory)

Reply:
The goal was to decouple the `SymbolicExecutionRunner` from the rules and let the factory create the analyzers, provide the supported rules and tell which are the ones enabled.

The `Rule` attributes on top are used only by the `RuleDescriptorGenerator` to generate the html description files. I cannot move yet those since the `RuleDetailBuilder.GetAllRuleDetails` looks only at the types which implement `DiagnosticAnalyser` to retrieve the rule ids: https://github.com/SonarSource/sonar-dotnet/blob/master/sonaranalyzer-dotnet/src/SonarAnalyzer.Utilities/RuleFinder.cs#L47.

I am more inclined to extend the `RuleFinder` to check also additional types and keep the responsibilities separated. What do you think?

Reply:
I've moved the Rule attributes in https://github.com/SonarSource/sonar-dotnet/pull/3338/commits/c79baa98e7fd7530d308ec45d9c5b36eb31a8e4d

Reply:
sounds good